### DISCUSSION TITLE
Option to normalize beatmap hitsounds

### DISCUSSION THREAD

**Discussion by SimplyMika** [ppy/osu#18714] [Ideas]
I would like there to be an option to normalize a beatmap's hitsound volumes.
Most maps don't ever use 100% hitsound volume, instead, they cap out at, say, 80% or even 60%, even in the loudest, most intense parts of the song. This can be a problem, because the ratio of song to hitsound volume can vary from map to map. Sometimes the hitsounds are painfully loud, sometimes you can barely hear them.

**How I picture this working:** Say you were playing a map with three different hitsound volumes – 10%, 50%, and 75%. If you had this setting turned on, they would instead behave as though they were 13.33%, 66.67%, and 100%. All the hitsound volumes get adjusted, so that the loudest one is at exactly 100%.

You might assume that mappers do a good job at matching the hitsound volume to the track volume, and often they do, but sometimes they really don't. This is pretty subjective, mappers might have different audio settings, or even different audio hardware which plays hitsounds (which tend to have higher frequencies) at a different volume.
Also, once the beatmap track volume gets normalized (as is planned, if I understand #1600 correctly), this will be important, because even if the mapper matched the volumes just right, all of a sudden the track volume is different to what the mapper heard (because it gets normalized) and now the hitsounds are too loud or too quiet.

You might also say that this should be handled through a rule in the ranking criteria, but I disagree. From my experience, the hitsound volume tends to be a lot worse the older a map is, so I think it's important to fix this retroactively, not just for new maps.

*Should this be a user setting?* I think so, but I would also be okay with a forced, game-wide implementation. I think once track volume is normalized, that would actually make a bunch of sense, but not right now.

**One issue I can see with this:** Perhaps a mapper starts to map a new beatmap, and sets the hitsound volume for the beginning part to something pretty quiet, say 20%. As they go to test-play the map, the hitsounds are very loud (100%), as they are the loudest in the entire map (because the map is unfinished).
Maybe this setting should not apply during test-play? Not sure if that's a good solution or not.

Thanks for reading :)

**Comment by smoogipoo** [comment]
Related: https://github.com/ppy/osu/discussions/18654. Maybe an option would be good.

**Comment by SevenChords** [comment]
As an addition to this:

I have encountered beatmaps (mostly in mania) where the dynamic range the beatmap hitsounds have is way too big to be comfortable (some hitsounds being not hearable at all and some blowing your eardrums out within the same beatmap).

An option for a minimum and a maximum hitsound volume level for the user would be great, to preserve hitsound dynamics intended by the beatmap author the hitsounds could be ~~normalized~~ scaled within the minimum/maximum hitsound volume set by the user.

**Comment by yamplum** [comment]
This would really help with maps that make certain hitsounds too quiet to provide feedback. For example, I've come across [a map](https://osu.ppy.sh/beatmapsets/1703935#osu/3481651) that mutes some slider end hits, making it hard to tell if I got it or not:
![image](https://user-images.githubusercontent.com/118788705/210289171-91c2ad5b-9686-44a2-8d00-f3a43c82e8c8.png)

**Reply by SimplyMika** [reply]
Note that the way I suggested this, that map wouldn't be affected much at all.
As you can see, there are 80% volume notes right next to the 5% ones. Assuming those are the loudest notes in the beatmap, that would mean that all hitsound volumes are increased by 20%, making 5% into 6%. Not a noticable difference.

I do think it could be useful to have something that deals with your issue there, but my suggestion doesn't.

**Reply by yamplum** [reply]
Yes, this is more to expand on @SevenChords's suggestion of setting the minimum levels as well. It should be fairly trivial to implement this as part of what you propose.
